lib/reline/line_editor.rb

```diff
@@ -406,109 +312,245 @@ def multiline_off
-        render_partial(prompt, prompt_width, line, @first_line_started_from)
-        move_cursor_down(@highest_in_all - (@first_line_started_from + @highest_in_this - 1) - 1)
-        scroll_down(1)
-        Reline::IOGate.move_cursor_column(0)
-        Reline::IOGate.erase_after_cursor
-      else
-        if not rendered and not @in_pasting
-          line = modify_lines(whole_lines)[@line_index]
-          prompt, prompt_width, prompt_list = check_multiline_prompt(whole_lines)
-          render_partial(prompt, prompt_width, line, @first_line_started_from)
-        end
-        render_dialog(cursor_column)
+    if old_levels.size > new_levels.size
+      Reline::IOGate.move_cursor_column new_levels.size
+      Reline::IOGate.erase_after_cursor
+    end
+  end
+
+  def editor_cursor_position
+    line = ' ' * calculate_width(prompt_list[@line_index], true) + whole_lines[@line_index].byteslice(0, @byte_pointer)
+    wrapped_line_before_cursor = split_by_width(line, screen_width).first.compact
+    editor_cursor_y = wrapped_lines[0...@line_index].sum(&:size) + wrapped_line_before_cursor.size - 1
+    editor_cursor_x = calculate_width(wrapped_line_before_cursor.last)
+    [editor_cursor_x, editor_cursor_y]
+  end
+
+  def clear_dialogs
+    @dialogs.each do |dialog|
+      dialog.contents = nil
+      dialog.trap_key = nil
+    end
+  end
+
+  def update_dialogs(key = nil)
+    @dialog_initialzed = true
+    editor_cursor_x, editor_cursor_y = editor_cursor_position
+    @dialogs.each do |dialog|
+      dialog.trap_key = nil
+      update_each_dialog(dialog, editor_cursor_x, editor_cursor_y - screen_scroll_top, key)
+    end
+  end
+
+  def clear_rendered_lines
+    Reline::IOGate.move_cursor_up @cursor_y
+    Reline::IOGate.move_cursor_column 0
+
+    num_lines = @rendered_screen_cache&.size
+    return unless num_lines && num_lines >= 1
+
+    Reline::IOGate.move_cursor_down num_lines - 1
+    (num_lines - 1).times do
+      Reline::IOGate.erase_after_cursor
+      Reline::IOGate.move_cursor_up 1
+    end
+    Reline::IOGate.erase_after_cursor
+    @rendered_screen_cache = nil
+  end
+
+  def render_full_content
+    lines = @buffer_of_lines.size.times.map do |i|
+      line = prompt_list[i] + modified_lines[i]
+      wrapped_lines, = split_by_width(line, screen_width)
+      wrapped_lines.last.empty? ? "#{line} " : line
+    end
+    @output.puts lines.map { |l| "#{l}\r\n" }.join
+  end
+
+  def print_nomultiline_prompt(prompt)
+    return unless prompt && !@is_multiline
+
+    # Readline's test `TestRelineAsReadline#test_readline` requires first output to be prompt, not cursor reset escape sequence.
+    @rendered_screen_cache = [[[0, Reline::Unicode.calculate_width(prompt, true), prompt]]]
+    @output.write prompt
+  end
+
+  def render_differential
+    unless @dialog_initialzed
+      update_dialogs
+    end
+
+    editor_cursor_x, editor_cursor_y = editor_cursor_position
+
+    rendered_lines = @rendered_screen_cache || []
```

Comment:
Among all 5 callers of `render_differential`, 4 of them sets `@rendered_screen_cache = nil` right before calling it. Only the one in `set_signal_handlers` doesn't do it.

That means for those 4 cases, `rendered_lines` is definitely `[]`, which maybe means the `if new_lines != rendered_lines` below is always true in those cases too 🤔 

If that's the case, we're essentially rerender everything in most cases rather than the differences?

Just want to bring it up to see if there's any room to further improve it.

Reply:
Adding `STDOUT.write "\a" unless @rendered_screen_cache` will make it easy to check by sound that when it isn't using the cache.

### L155 in `resize`
Resizing window will word wrap rendered text. This will make it impossible to reuse the screen. Need to rerender everything.

### L161 in `set_signal_handlers`
Press C-c will clear dialogs before aborting the input.
This will render the diffence in dialog position.

### L530 in `rerender_all`
### L553 in `rerender`
This is the most frequently called method from reline.rb.
Most of the cases, it uses non-nil value of `@rendered_screen_cache`.
This method is called with `@rendered_screen_cache = nil` in a few cases
- Just after IRB startup
- Just after previous input is finished and rendering the next prompt
- Just after input is interrupted with C-c and rendering the next prompt

Maybe this two methods can be combined into one.

### L543 in `handle_cleared`
When screen is cleared with `C-l`. Need to rerender everything.

Reply:
Combined `rerender_all` and `rerender` into `def rerender`, simplified rerender call from reline.rb in https://github.com/ruby/reline/pull/614/commits/961c294ca07cc614e2a60d75563441827d139503